build_runner_core/lib/src/package_graph/package_graph.dart

```diff
@@ -78,14 +78,31 @@ class PackageGraph {
           'pubspec.yaml.');
     }
 
-    final packageConfig =
-        await findPackageConfig(Directory(packagePath), recurse: false);
+    // The path of the directory that contains .dart_tool/package_config.json.
```

Comment:
Is this different from just passing `recurse: true`?

Reply:
The only difference is that we can use the discovered path `rootDir ` where we find the package config below.

With `recurse: true` that would not be the case - we would only be given the packageConfig itself...

Reply:
Sure, we could just set the root dir as the parent of the package config dir though. Not a big difference either way but I would slightly prefer it that way.

On a related note though - I am not sure we actually want to load the pubspec from the workspace root instead of the package root? We will end up seeing all the deps used by any package in the workspace then, instead of just the deps of the current package. Unless the current package doesn't list its deps at all?

Reply:
The only way I can see to do that would be to parse pubspec.yaml of the current package and all its dependencies transitively.

Reply:
> Sure, we could just set the root dir as the parent of the package config dir

Problem is, that with `findPackageConfig` we don't get a package config dir. Only the package config **object** (that doesn't expose it's location if I understand right).

Reply:
There are also likely to be some ordering issues with build_runner here, we won't know the correct dependency ordering for packages within the workspace, or where cycles exist. That won't always matter but it could sometimes, especially if builders are defined within the package.

Reply:
Sorry I hadn't seen your other comments when I did my last response :).

But in any case:

> I'm not sure how build runner uses this dependency graph - we might need to prune it?

Some builders will generate code for all packages in the whole dependency graph, so we could do extra work here if we see deps that are never used.

> The only way I can see to do that would be to parse pubspec.yaml of the current package and all its dependencies transitively.

Right, this is what we do today, we parse all transitive pubspec.yaml files. We want to know the entire dependency structure including cycles, as it influences the order in which we build things.

> Problem is, that with findPackageConfig we don't get a package config dir. Only the package config object (that doesn't expose it's location if I understand right).

Oh, ok then this code makes sense. Is there a different API that just gives us the location?

> Is this a new problem? Did we know the ordering before?

Answered above, we crawl transitive pubspecs.

Reply:
> Answered above, we crawl transitive pubspecs.

Ah - then there should be no problem if we just start from the pubspec in `packagePath` (which we still do with this change) the workspace-wide package config and lockfile will just contain a superset of the needed packages.

> Is there a different API that just gives us the location?

I don't think so. That's why I wrote it.

Reply:
Ah ok, if we are still reading the actual package pubspec and not a root one, this is probably all fine. As long as we handle the case where there are more packages in the package config than we can find in the pubspecs.

Reply:
If I read this code right: https://github.com/dart-lang/build/blob/db5180c903b27fd1c3d344c5b22c396540ea0c78/build_runner_core/lib/src/package_graph/package_graph.dart#L133-L144

The package graph will contain all packages in the workspace-wide resolution, but will have the correct root. And if I'm not wrong, only the reachable packages are handled according to:

https://github.com/dart-lang/build/blob/5eb5d9bf419b43252c7f1a41d47f334ab675e654/build_runner/lib/src/build_script_generate/build_script_generate.dart#L82